### PR TITLE
sock/async: add function to retrieve session object of current DTLS event

### LINKED ISSUE #15517: pkg/tinydtls: async event SOCK_ASYNC_CONN_FIN does not indicate which session has been finished
#### Description

When using tinyDTLS and the sock/dtls API with asynchronous sockets the event for a finished session does not indicate which session has been closed. It makes it impossible to keep the session for a long time because you cannot track sessions properly without touching the internals of tinydtls.

The following code block is the part in sock_dtls.c which handles this part. The internal event gets the correct closed session it is just not passed to the callback. I'm not quite sure what `sock->async_cb(sock, SOCK_ASYNC_CONN_FIN, sock->async_cb_arg);` hands over to the cb function. sock->async_cb_arg? Whats that? 
```
pkg/tinydtls/contrib/sock_dtls.c

static int _event(struct dtls_context_t *ctx, session_t *session,
                  dtls_alert_level_t level, unsigned short code)
{
    (void)level;
    (void)session;

    sock_dtls_t *sock = dtls_get_app_data(ctx);
    msg_t msg = { .type = code, .content.ptr = session };

    if (IS_ACTIVE(ENABLE_DEBUG)) {
        switch (code) {
            case DTLS_EVENT_CONNECT:
                DEBUG("sock_dtls: event connect\n");
                break;
            case DTLS_EVENT_CONNECTED:
                DEBUG("sock_dtls: event connected\n");
                break;
            case DTLS_EVENT_RENEGOTIATE:
                DEBUG("sock_dtls: event renegotiate\n");
                break;
        }
    }
    if (!level && (code != DTLS_EVENT_CONNECT)) {
        mbox_put(&sock->mbox, &msg);
    }
#ifdef SOCK_HAS_ASYNC
    if (sock->async_cb != NULL) {
        switch (code) {
            case DTLS_ALERT_CLOSE_NOTIFY:
                /* peer closed their session */
                if(session) {
                    printf("session is != null!\n");
                    printf("Port: %d\n", session->port);
                }
                sock->async_cb(sock, SOCK_ASYNC_CONN_FIN, sock->async_cb_arg);
                break;
            case DTLS_EVENT_CONNECTED:
                /* we received a session handshake initialization */
                sock->async_cb(sock, SOCK_ASYNC_CONN_RECV,
                               sock->async_cb_arg);
                break;
            default:
                break;
        }
    } else {
        printf("not existing\n");
    }
#endif
    return 0;
}
```

#### Expected results
Argument of the callback event points to the closed session.

#### Actual results
The argument is currently null.

### PR DESCRIPTION
**Pull request by janosbrodbeck**
### Contribution description
This PR adds a new function to sock async: `sock_dtls_get_event_session()` to retrieve the session object within a DTLS event. Currently it is not possible to get this session for every event type (e.g. SOCK_ASYNC_CONN_FIN, SOCK_ASYNC_CONN_RDY). It should work for every possible event type. The issue was described in #15517. 

I didn't get it to work properly with just a pointer in the socket structure, since the session needed to be stored somewhere and storing it was not possible within the `_event()` function of tinydtls. So I placed it right away in the socket. 
 
Not sure whether I placed the new function correctly. 

### Testing procedure
Implementation in tinyDTLS can be tested with [`pkg_tinydtls_sock_async`](https://github.com/RIOT-OS/RIOT/tree/master/tests/pkg_tinydtls_sock_async)

* Flash and run on two devices (e.g. native)
* Execute dtlss on one of the devices
* Execute dtlsc {ip of other node} {some data}  on the other device

### Issues/PRs references
Fixes  #15517

### CODE REVIEW

**Comment by janosbrodbeck** [comment]
I have also added a very small doc fix commit. The documentation of `sock_dtls_set_cb` wrote something about getting something but it just sets stuff. So I've changed Get to Set. But can drop this commit if unwanted or wrong.

File: pkg/tinydtls/contrib/sock_dtls.c [hunk 1]
```diff
@@ -87,6 +87,7 @@ static int _read(struct dtls_context_t *ctx, session_t *session, uint8_t *buf,
     sock->buffer.session = session;
 #ifdef SOCK_HAS_ASYNC
     if (sock->async_cb != NULL) {
+        memcpy(&sock->async_cb_session, session, sizeof(session_t));
```
Comment (pokgak) [inline pkg/tinydtls/contrib/sock_dtls.c:90]: The `memcpy` seems unnecessary if the sessions are going to be used in the handler anyways because the handler is called directly after by `sock->async_cb`, so we don't have to worry about the memory of `session` going out of scope. A pointer would be enough. 

If the server wants to keep that information session after the handler exits, only then do the session need to be copied but that should be done by the application (dtls-client/server) and not `sock_dtls`.

Comment (pokgak) [reply pkg/tinydtls/contrib/sock_dtls.c:90]: Though by not copying the session into the sock, `sock_dtls_get_event_session` would only be usable withing the async handler, as outside of it there is no guarantee that the pointer is valid and therefore should set to `NULL` when exiting the handler. This behavior should still solve the issue you're having in #15517 where you cannot access the session from the handler. Is there any uses for `sock_dtls_get_event_session` outside of the handler?

File: pkg/tinydtls/include/sock_dtls_types.h [hunk 2]
```diff
@@ -34,6 +34,13 @@ extern "C" {
 #define SOCK_DTLS_MBOX_SIZE     (4)         /**< Size of DTLS sock mailbox */
 #endif
 
+/**
+ * @brief Information about remote client connected to the server
+ */
+struct sock_dtls_session {
+    session_t       dtls_session;    /**< TinyDTLS session */
+};
```
Comment (miri64) [inline pkg/tinydtls/include/sock_dtls_types.h:42]: Why move this?

Comment (janosbrodbeck) [reply pkg/tinydtls/include/sock_dtls_types.h:42]: Oh, just a leftover from another approach I tried. Will fix!

File: pkg/tinydtls/contrib/sock_dtls.c (hunk 1 above)
Comment (janosbrodbeck) [reply pkg/tinydtls/contrib/sock_dtls.c:90]: The problem I had with a pointer was that I cannot set the pointer to NULL after the callback call.
```
sock->async_cb_session = session;
sock->async_cb(sock, SOCK_ASYNC_MSG_RECV, sock->async_cb_arg);
sock->async_cb_session = NULL;
```
The session is already NULL again when I call `sock_dtls_get_event_session` within the DTLS handler. 

No, `sock_dtls_get_event_session` should have no use case outside of the handler. It's not even guaranteed that you get a proper information out of it outside the handler.

Comment (janosbrodbeck) [reply pkg/tinydtls/contrib/sock_dtls.c:90]: I have a branch with the pointer solution [here](https://github.com/janosbrodbeck/RIOT/tree/pr/dtls/get_event_session_pointer). I would prefer a solution with just a pointer. 

Or did I misunderstood you? :)

Comment (pokgak) [reply pkg/tinydtls/contrib/sock_dtls.c:90]: the code in your branch is exactly how I imagined it. Weird that the session is already NULL though..

Comment (pokgak) [reply pkg/tinydtls/contrib/sock_dtls.c:90]: Okay apparently I misunderstood how `sock_async` works. Although it seems that `sock->async_cb` is called inside of `_event`/`_read`, it is actually executed after leaving those functions. I discovered this after adding some printf. Can you confirm this @miri64?

This means, we accessed the session after we set it to `NULL`, causing it to segfault. So I tested without setting `async_cb_session` to `NULL` and it worked! Though I'm not sure if this have any other side effects..

Here is the log of the working test with more printf added:
```
main(): This is RIOT! (Version: 2021.01-devel-1872-gfdb01-pr/dtls/get_event_session_pointer)                                                                                                                       
DTLS sock example application                                                                                                                                                                                      
All up, running the shell now                                                                                                                                                                                      
> dtlss start                                                                                                                                                                                                      
dtlss start                                                                                                                                                                                                        
> entering _event                                                                                                                                                                                                  
leaving _event                                                                                                                                                                                                     
entering _dtls_handler                                                                                                                                                                                             
event: SOCK_ASYNC_CONN_RECV
Session handshake received: Peer: FE:80:00:00:00:00:00:00:E8:BC:C0:FF:FE:06:47:8B: Port: 12345 Family: 4
New client connected
leaving _dtls_handler
entering _dtls_handler
event: SOCK_ASYNC_CONN_RDY
Session became ready: Peer: FE:80:00:00:00:00:00:00:E8:BC:C0:FF:FE:06:47:8B: Port: 12345 Family: 4
leaving _dtls_handler
entering _read
leaving _read
entering _dtls_handler
event: SOCK_ASYNC_MSG_RECV
Message received: done mesg recevd
Peer: FE:80:00:00:00:00:00:00:E8:BC:C0:FF:FE:06:47:8B: Port: 12345 Family: 4
Received 12 bytes -- (echo)
leaving _dtls_handler
entering _dtls_handler
leaving _dtls_handler
entering _event
leaving _event
entering _dtls_handler
event: SOCK_ASYNC_CONN_FIN
Session was destroyed by peer: Peer: FE:80:00:00:00:00:00:00:E8:BC:C0:FF:FE:06:47:8B: Port: 12345 Family: 4

leaving _dtls_handler

```

**Comment by pokgak** [comment]
> (Is there something like a spoiler or collapse tag?)

You can use the HTML `<details></details>` tag: 

<details>HIDE ME</details>

File: pkg/tinydtls/contrib/sock_dtls.c (hunk 1 above)
Comment (janosbrodbeck) [reply pkg/tinydtls/contrib/sock_dtls.c:90]: Ahh was wondering too when this gets called.
The problem I saw with not setting it to NULL is that it points to "random" memory and could potentially be misused?

Comment (miri64) [reply pkg/tinydtls/contrib/sock_dtls.c:90]: > Okay apparently I misunderstood how `sock_async` works. Although it seems that `sock->async_cb` is called inside of `_event`/`_read`, it is actually executed after leaving those functions. I discovered this after adding some printf. Can you confirm this @miri64?

Mhh there are actually two callbacks, that's maybe where the confusion is coming from. `sock->async_cb` is the asynchronous callback internal to the `sock_async` backend and then there is the frontend's callback (since `sock_async_event` also uses callbacks) which is called from the event handler (to make things portable between different frontend approaches including a bare metal approach of just using the backend they have the same signature). `sock->async_cb` is called from within `_event` and `_read` that is correct, but it merely triggers the frontend event (`event_post()` in the `sock_async_event` case). The actual event handling is done within the the event handler of the frontend, which will be (comparably) much later then the call to `sock->async_cb` and in a different thread (the medium priority thread of `event_thread` in `tests/pkg_tinydtls_sock_async`'s case. Hope that clears things

TL;DR passing pointers around breaks the asynchronous and decoupled nature of `sock_async`. @janosbrodbeck was right in copying the session instead.

Comment (miri64) [reply pkg/tinydtls/contrib/sock_dtls.c:90]: Question is: what do you do, if there were multiple events fired before the handler comes to handling it. With your approach you will always only have the chance to get the session of the last event fired (which given that events are identified by a bitfield you can't determined which it was).

Comment (janosbrodbeck) [reply pkg/tinydtls/contrib/sock_dtls.c:90]: Thanks for the explanation. That makes the process more understandable. Will have to look at this mechanism in detail the next days...that's something you should always have in mind when using async callbacks.

> Question is: what do you do, if there were multiple events fired before the handler comes to handling it. With your approach you will always only have the chance to get the session of the last event fired (which given that events are identified by a bitfield you can't determined which it was).

I also thought of this while reading. The approach here then no longer works. The other idea with another parameter in the callback does not work then either. Hmm...

Comment (miri64) [reply pkg/tinydtls/contrib/sock_dtls.c:90]: This is about getting the session for very dedicated event types like `SOCK_ASYNC_CONN_FIN` or `SOCK_ASYNC_CONN_RDY`, right? Those are pretty rare and won't probably fire twice on the same `sock`. How about only setting the sessions for occurences of that event and specifically document that. For `SOCK_ASYNC_MSG_RECV` a user should use `sock_dtls_recv` anyway.

Comment (janosbrodbeck) [reply pkg/tinydtls/contrib/sock_dtls.c:90]: Only those two events, yes. The number of sessions that a device can hold will probably never be huge. Sessions are expensive (even more so with wolfSSL). I also think it is very unlikely that this scenario will occur in practice. But i also don't want to exclude it to be honest. However, when it happens you can actually prevent resulting problems at the application level. It's not optimal...but I don't know any other solution at the moment. 

In case of doubt if problems occur we could make setting both event types configurable. Which means: If you only need one event type you could disable the setting of the other event type. This should lower the chance of the problem even more since then the session will be set for only one event type. For example I don't need the `SOCK_ASYNC_CONN_RDY` for gcoap currently.

**Comment by janosbrodbeck** [comment]
Pushed the discussed changes. Now, the event session is only set for `SOCK_ASYNC_CONN_READY` and `SOCK_ASYNC_CONN_FIN`. Added this info into the note section of the function documentation. Thought about putting this info into a warning section, but that may be a bit overkill. 

Did not adjust the provided test procedure. You can see the expected behaviour in the test. For the `SOCK_ASYNC_CONN_RECV` event the retrieved session is still the previously set session or empty if it's the first.

**Comment by janosbrodbeck** [comment]
Also adjusted the line length of the function head to fit into the 80 character recommendation.

**Comment by pokgak** [comment]
LGTM. Can you also include the patch (only CONN_RECV and CONN_FIN part) as commit to show usage of `sock_dtls_get_event_session` in the DTLS handler so that it will at least get tested the next time someone run the test?

File: sys/include/net/sock/async.h [hunk 3]
```diff
@@ -33,7 +33,22 @@ extern "C" {
...
 
 /**
- * @brief   Gets the asynchronous event context from sock object
+ * @brief Gets the asynchronous event session from sock object
+ *
+ * @pre `(sock != NULL) && (session != NULL)`
+ *
+ * @note Only available with @ref SOCK_HAS_ASYNC defined.
+ *       Should only be called within a DTLS event and only for the event
+ *       types @ref SOCK_ASYNC_CONN_RDY and @ref SOCK_ASYNC_CONN_FIN.
```
Comment (miri64) [inline sys/include/net/sock/async.h:42]: Mhhh maybe add a return value, that signals to the caller when the the session is available or not? I know this makes the implementation more complicated than it is now, but `sock` is an end-user API that is supposed to be easy to use. Could be boolean or an `errno`-based status code

**Review by miri64 — CHANGES_REQUESTED** [review]

File: sys/include/net/sock/async.h (hunk 3 above)
Comment (janosbrodbeck) [reply sys/include/net/sock/async.h:42]: Totally forgot, sorry. Squashed the undo directly with the related commit. 

Regarding the return value. It would be nice to have it...but how?  The only idea I have for this currently is to also add (e.g.) a boolean to the socket object and set this boolean when a event occurs. So setting it for the CONN_RDY, CONN_FIN to available and for every other event to not available. Would that be sufficient or do you imagine it differently?

Comment (miri64) [reply sys/include/net/sock/async.h:42]: In the specific case of `tinydtls` you could use [the `size` member of the `session_t` struct](https://github.com/eclipse/tinydtls/blob/706888256c3e03d9fcf1ec37bb1dd6499213be3c/session.h#L40-L49) as a signifier if the session is set.

```C
typedef struct {
  socklen_t size;		/**< size of addr */
  union {
    struct sockaddr     sa;
    struct sockaddr_storage st;
    struct sockaddr_in  sin;
    struct sockaddr_in6 sin6;
  } addr;
  uint8_t ifindex;
} session_t;
```

The object needs to zero'd of course. However, I would do this on read and socket init. This way, if you e.g. receive a CONN_RDY and a MSG_RECV in the same callback you don't accidentally remove the session before it can be read.

Comment (janosbrodbeck) [reply sys/include/net/sock/async.h]: Done :) Thank you! I've adjusted the test procedure to test, that the session is not available for other event types.

**Comment by janosbrodbeck** [comment]
> LGTM. Can you also include the patch (only CONN_RECV and CONN_FIN part) as commit to show usage of `sock_dtls_get_event_session` in the DTLS handler so that it will at least get tested the next time someone run the test?

I've added the usage of the new function to the test. Are you okay with it the way it is now?

File: tests/pkg_tinydtls_sock_async/dtls-server.c [hunk 4]
```diff
@@ -117,10 +117,41 @@ static void _dtls_handler(sock_dtls_t *sock, sock_async_flags_t type, void *arg)
         puts("New client connected");
     }
     if (type & SOCK_ASYNC_CONN_FIN) {
-        puts("Session was destroyed by peer");
+        if (sock_dtls_get_event_session(sock, &session))
+        {
```
Comment (pokgak) [inline tests/pkg_tinydtls_sock_async/dtls-server.c:121]: minor nit: put the braces on the same line as the if statement

**Review by pokgak — APPROVED** [review]

File: tests/pkg_tinydtls_sock_async/dtls-server.c (hunk 4 above)
Comment (janosbrodbeck) [reply tests/pkg_tinydtls_sock_async/dtls-server.c]: fixed! And also uncluttered the DTLS handler now

File: tests/pkg_tinydtls_sock_async/dtls-server.c [hunk 5]
```diff
@@ -105,9 +106,9 @@ static void _dtls_handler(sock_dtls_t *sock, sock_async_flags_t type, void *arg)
         return;
     }
 
-    if (type & SOCK_ASYNC_CONN_RECV) {
-        sock_dtls_session_t session = { 0 };
+    sock_dtls_session_t session = { 0 };
 
+    if (type & SOCK_ASYNC_CONN_RECV) {
         puts("Session handshake received");
```
Comment (miri64) [inline tests/pkg_tinydtls_sock_async/dtls-server.c:112]: Here and in the other event types that are not `SOCK_ASYNC_CONN_FIN` or `SOCK_ASYNC_CONN_RDY` (as well in the client), you could also use `expect(!sock_dtls_get_event_session(sock, &session))` to make this proper tests ;-).

Comment (janosbrodbeck) [reply tests/pkg_tinydtls_sock_async/dtls-server.c:115]: Done! In the event types `SOCK_ASYNC_CONN_FIN` and `SOCK_ASYNC_CONN_RDY` we could also use `expect(sock_dtls....)` to test it since currently it is only printed and not tested. Would that be too much to also add it there? Would like to keep the session printing, because it is currently the only example usage of this new function.

Comment (miri64) [reply tests/pkg_tinydtls_sock_async/dtls-server.c:115]: > Would like to keep the session printing, because it is currently the only example usage of this new function.

Yes, as you print output, that provides enough result (more than expect which does not print anything when succeeding)

File: tests/pkg_tinydtls_sock_async/dtls-client.c [hunk 6]
```diff
@@ -134,9 +138,19 @@ static void _dtls_handler(sock_dtls_t *sock, sock_async_flags_t type, void *arg)
         _close_sock(sock);
     }
     if (type & SOCK_ASYNC_CONN_RDY) {
-        puts("Session became ready");
+        if (sock_dtls_get_event_session(sock, &session)) {
+            sock_udp_ep_t ep;
+            sock_dtls_session_get_udp_ep(&session, &ep);
+
+            printf("Session became ready: ");
+            print_ipv6_addr_with_port(ep.addr.ipv6, sizeof(ep.addr), ep.port);
```
Comment (miri64) [inline tests/pkg_tinydtls_sock_async/dtls-client.c:146]: Any reason why you not use `sock_udp_ep_fmt()` from `sys/include/net/sock/util.h` here? Seems like re-inventing the wheel.

**Comment by miri64** [comment]
One last comment. Other than that: I tested everything and am fine with most changes (apart from the one mentioned above). You may squash directly.

File: tests/pkg_tinydtls_sock_async/dtls-client.c (hunk 6 above)
Comment (janosbrodbeck) [reply tests/pkg_tinydtls_sock_async/dtls-client.c:146]: I just knew the format function when the address is a `ipv6_addr_t` object but didn't see one for an endpoint. Was not my intention to reinvent, just did not found a proper function.

**Comment by janosbrodbeck** [comment]
Addressed last comment and squashed. Test uses `sock_udp_ep_fmt()` now for printing.

**Review by miri64 — APPROVED** [review]
ACK. Re-tested and everything works as expected. :-)

**Comment by janosbrodbeck** [comment]
Thanks for the help and review @miri64 and @pokgak !
